### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by GuilhE** [TakuSemba/Spotlight#93]
This is how I setup the coach mark:

  ```
private lateinit var spotlight : Spotlight

private fun showCoachMarksForHome() {
           spotlight = Spotlight.Builder(this)
                    ...
                    .setOnSpotlightListener(object : OnSpotlightListener { ... } {
                        override fun onStarted()  { ... }
                        override fun onEnded() { ... }
                    })
                    .build()
           spotlight.start()
}
```

And when the Activity is destroyed I get:

>     HEAP ANALYSIS RESULT
>     ====================================
>     1 APPLICATION LEAKS
>     
>     References underlined with "~~~" are likely causes.
>     Learn more at https://squ.re/leaks.
>     
>     1556569 bytes retained by leaking objects
>     Signature: cb8204dece4c8b1d5a57e81e99d939e5467d1d7
>     ┬───
>     │ GC Root: System class
>     │
>     ├─ android.os.Looper class
>     │    Leaking: NO (Thread↓ is not leaking and a class is never leaking)
>     │    ↓ static Looper.sMainLooper
>     ├─ android.os.Looper instance
>     │    Leaking: NO (Thread↓ is not leaking)
>     │    ↓ Looper.mThread
>     ├─ java.lang.Thread instance
>     │    Leaking: NO (the main thread always runs)
>     │    Thread name: 'main'
>     │    ↓ Thread.threadLocals
>     │             ~~~~~~~~~~~~
>     ├─ java.lang.ThreadLocal$ThreadLocalMap instance
>     │    Leaking: UNKNOWN
>     │    ↓ ThreadLocal$ThreadLocalMap.table
>     │                                 ~~~~~
>     ├─ java.lang.ThreadLocal$ThreadLocalMap$Entry[] array
>     │    Leaking: UNKNOWN
>     │    ↓ ThreadLocal$ThreadLocalMap$Entry[].[22]
>     │                                         ~~~~
>     ├─ java.lang.ThreadLocal$ThreadLocalMap$Entry instance
>     │    Leaking: UNKNOWN
>     │    ↓ ThreadLocal$ThreadLocalMap$Entry.value
>     │                                       ~~~~~
>     ├─ android.animation.AnimationHandler instance
>     │    Leaking: UNKNOWN
>     │    ↓ AnimationHandler.mAnimationCallbacks
>     │                       ~~~~~~~~~~~~~~~~~~~
>     ├─ java.util.ArrayList instance
>     │    Leaking: UNKNOWN
>     │    ↓ ArrayList.elementData
>     │                ~~~~~~~~~~~
>     ├─ java.lang.Object[] array
>     │    Leaking: UNKNOWN
>     │    ↓ Object[].[0]
>     │               ~~~
>     ├─ android.animation.ValueAnimator instance
>     │    Leaking: UNKNOWN
>     │    ↓ ValueAnimator.mUpdateListeners
>     │                    ~~~~~~~~~~~~~~~~
>     ├─ java.util.ArrayList instance
>     │    Leaking: UNKNOWN
>     │    ↓ ArrayList.elementData
>     │                ~~~~~~~~~~~
>     ├─ java.lang.Object[] array
>     │    Leaking: UNKNOWN
>     │    ↓ Object[].[0]
>     │               ~~~
>     ├─ com.takusemba.spotlight.SpotlightView$invalidator$1 instance
>     │    Leaking: UNKNOWN
>     │    Anonymous class implementing android.animation.ValueAnimator$AnimatorUpdateListener
>     │    ↓ SpotlightView$invalidator$1.this$0
>     │                                  ~~~~~~
>     ├─ com.takusemba.spotlight.SpotlightView instance
>     │    Leaking: YES (View.mContext references a destroyed activity)
>     │    mContext instance of ...HomeActivity with mDestroyed = true
>     │    View#mParent is null
>     │    View#mAttachInfo is null (view detached)
>     │    View.mWindowAttachCount = 1
>     │    ↓ SpotlightView.mContext
>     ╰→ ....HomeActivity instance
>     ​     Leaking: YES (ObjectWatcher was watching this because ...HomeActivity received Activity#onDestroy() callback and Activity#mDestroyed is true)
>     ​     key = ad3de779-b9ee-4a66-9a94-444adc836ee2
>     ​     watchDurationMillis = 5217
>     ​     retainedDurationMillis = 217
>     ====================================
>     0 LIBRARY LEAKS
>     
>     A Library Leak is a leak caused by a known bug in 3rd party code that you do not have control over.
>     See https://square.github.io/leakcanary/fundamentals-how-leakcanary-works/#4-categorizing-leaks
>     ====================================
>     METADATA
>     
>     Please include this in bug reports and Stack Overflow questions.
>     
>     Build.VERSION.SDK_INT: 29
>     Build.MANUFACTURER: Google
>     LeakCanary version: 2.4
>     Analysis duration: 17257 ms
>     Heap dump timestamp: 1596552647400
>     ====================================

Any hint?
In the meanwhile I'll try to fix this and open a PR.

Thanks.

**Comment by TakuSemba** [comment]
Hi, thank you for the issue.

I confirmed that SpotlightView is leaked from an unremoved listener.
I will fix it in the next version.

I appreciate your help.

**Comment by GuilhE** [comment]
> Hi, thank you for the issue.
> 
> I confirmed that SpotlightView is leaked from an unremoved listener.
> I will fix it in the next version.
> 
> I appreciate your help.

I've tried to use WeakReferences for the callbacks but unfortunately didn't work. I didn't try replacing with Higher-Order Functions though.

**Comment by TakuSemba** [comment]
I believe I fixed it on v2.0.2.

Please reopen this if this still happens.
